Address Clang-Tidy diagnostics

jngrad opened the pull request:
Address various diagnostics emitted by Clang-Tidy with Clang 19 and IntelLLVM 2024.2.1. Use `NOLINT` where justified.

dalg24 commented on .clang-tidy:
```diff
@@ -13,7 +13,15 @@ Checks: >
   modernize-type-traits,
   modernize-use-nullptr,
   modernize-use-using,
+  modernize-use-override,
+  modernize-use-equals-delete,
+  modernize-use-equals-default,
+  modernize-use-bool-literals,
+  modernize-loop-convert,
```
Do not attempt to enable more checks as part of this PR.
It is unclear that we want to enable them and we would only treat them one by one.

jngrad replied:
I've reverted these changes.

dalg24 reviewed (changes requested):
Please focus on checks that are currently enabled and that yield warnings with newer toolchains.
You are welcome to propose the typo fixes as a separate PR.

jngrad committed:
Address Clang-Tidy diagnostics

clang-analyzer-security.insecureAPI.strcpy
clang-diagnostic-float-conversion
clang-diagnostic-implicit-int-float-conversion
bugprone-multi-level-implicit-pointer-conversion
bugprone-return-const-ref-from-parameter
bugprone-crtp-constructor-accessibility
modernize-loop-convert
modernize-use-nullptr
modernize-use-override
modernize-use-equals-delete
modernize-use-equals-default
modernize-use-bool-literals
readability-non-const-parameter
readability-simplify-boolean-expr

jngrad commented:
> Please focus on checks that are currently enabled and that yield warnings with newer toolchains.

I've amended the PR accordingly.

> You are welcome to propose the typo fixes as a separate PR.

Good idea.
